test/path-or-subdomain.spec.ts

```diff
@@ -1,49 +1,169 @@
+        href: `http://example.com/ipfs/${cid}/foo/bar.txt`
+      })
+
+      const out = toSubdomainRequest(loc)
+      const exp = new URL(`http://${cidV1}.ipfs.example.com/`)
+      exp.searchParams.set('helia-sw', '/foo/bar.txt')
+
+      expect(out).to.equal(exp.href)
+    })
+
+    it('preserves existing ?search=params and #hash', () => {
+      const cid = 'QmbQDovX7wRe9ek7u6QXe9zgCXkTzoUSsTFJEkrYV1HrVR'
+      const cidV1 = CID.parse(cid).toV1().toString(base32)
+      const loc = makeLoc({
+        pathname: `/ipfs/${cid}/path/to/file`,
+        search: '?foo=bar&baz=qux',
+        hash: '#section2',
+        href: `https://example.com/ipfs/${cid}/path/to/file?foo=bar&baz=qux#section2`,
+        protocol: 'https:'
+      })
+
+      const out = toSubdomainRequest(loc)
+      const exp = new URL(`https://${cidV1}.ipfs.example.com/`)
+      exp.searchParams.set('helia-sw', '/path/to/file')
+      exp.searchParams.set('foo', 'bar')
+      exp.searchParams.set('baz', 'qux')
+      exp.hash = '#section2'
+
+      expect(out).to.equal(exp.href)
+    })
+
+    it('drops helia-sw when there is no “extra” path (i.e. only /ipfs/<cid>)', () => {
+      const cid = 'QmbQDovX7wRe9ek7u6QXe9zgCXkTzoUSsTFJEkrYV1HrVR'
+      const cidV1 = CID.parse(cid).toV1().toString(base32)
+      const loc = makeLoc({
+        pathname: `/ipfs/${cid}`,
+        href: `http://example.com/ipfs/${cid}`
+      })
+
+      const out = toSubdomainRequest(loc)
+      expect(out).to.equal(`http://${cidV1}.ipfs.example.com/`)
+    })
+
+    it('handles /ipns/<libp2p-key>/… by converting to Base36 and preserving rest', () => {
+      // Use a valid libp2p-key CIDv1 in base36 for testing
+      const key = 'k51qzi5uqu5dh9ihj4p2v5sl3hxvv27ryx2w0xrsv6jmmqi91t9xp8p9kaipc2'
+      const keyV1 = CID.parse(key).toV1().toString(base36)
+      const loc = makeLoc({
+        protocol: 'https:',
+        host: 'gateway.local',
+        pathname: `/ipns/${key}/blog/post`,
+        href: `https://gateway.local/ipns/${key}/blog/post`
+      })
+
+      const out = toSubdomainRequest(loc)
+      const exp = new URL(`https://${keyV1}.ipns.gateway.local/`)
+      exp.searchParams.set('helia-sw', '/blog/post')
+      expect(out).to.equal(exp.href)
+    })
+
+    it('falls back to dnsLink-encoding when the second segment is not a CID but contains dots', () => {
+      const hostname = 'mysite.local'
+      const loc = makeLoc({
+        host: hostname,
+        pathname: '/ipns/foo.bar/baz',
+        href: `http://${hostname}/ipns/foo.bar/baz`
+      })
+
+      const out = toSubdomainRequest(loc)
+      const url = new URL(out)
+      expect(url.origin).to.equal(`http://foo-bar.ipns.${hostname}`)
+      expect(url.pathname).to.equal('/')
+      expect(url.searchParams.get('helia-sw')).to.equal('/baz')
+    })
+
+    it('ignores invalid namespaces', () => {
+      // TODO: This test was added without modifying the code, and i'm not sure we want this functionality.
```

Comment:
callout for attention here. there are checks and guards elsewhere that ensure users aren't being redirected to this actual domain if its not ipfs/ipns.. but loading https://inbrowser.dev/potato/QmWhatever/foo results in the landing page, and we should probably have some error screen for this